pkg/apis/serving/v1beta1/inference_service_defaults.go

```diff
@@ -384,15 +276,32 @@ func (isvc *InferenceService) setMlServerDefaults() {
 	}
 }
 
-func (isvc *InferenceService) setTorchServeDefaults() {
+func (isvc *InferenceService) SetTorchServeDefaults() {
+	// set 'v1' as default protocol version for torchserve
+	if isvc.Spec.Predictor.Model.ProtocolVersion == nil {
+		protocolV1 := constants.ProtocolV1
+		isvc.Spec.Predictor.Model.ProtocolVersion = &protocolV1
+	}
 	// set torchserve service envelope based on protocol version
 	if isvc.ObjectMeta.Labels == nil {
 		isvc.ObjectMeta.Labels = map[string]string{constants.ServiceEnvelope: constants.ServiceEnvelopeKServe}
 	} else {
 		isvc.ObjectMeta.Labels[constants.ServiceEnvelope] = constants.ServiceEnvelopeKServe
 	}
-	if isvc.Spec.Predictor.Model.ProtocolVersion != nil &&
-		constants.ProtocolV2 == *isvc.Spec.Predictor.Model.ProtocolVersion {
+	if constants.ProtocolV2 == *isvc.Spec.Predictor.Model.ProtocolVersion {
 		isvc.ObjectMeta.Labels[constants.ServiceEnvelope] = constants.ServiceEnvelopeKServeV2
 	}
 }
+
+func (isvc *InferenceService) SetTritonDefaults() {
+	// set 'v2' as default protocol version for triton server
+	if isvc.Spec.Predictor.Model.ProtocolVersion == nil {
+		protocolV2 := constants.ProtocolV2
+		isvc.Spec.Predictor.Model.ProtocolVersion = &protocolV2
+	}
+	// set model-control-model arg to 'explicit' if storage uri is nil
+	if isvc.Spec.Predictor.Model.StorageURI == nil {
+		isvc.Spec.Predictor.Model.Args = append(isvc.Spec.Predictor.Model.Args,
+			fmt.Sprintf("%s=%s", "--model-control-mode", "explicit"))
+	}
```

Comment:
wondering if this can be moved to the triton serving runtime itself instead of explicit code here.

Reply:
@yuzisun If we add `--model-control-mode=explicit` arg for single model serving(storageUri != nil), the model is not registered.

Reply:
@Suresh-Nakkeran @yuzisun I'm also wondering if we could handle most of this other stuff in the serving runtimes. Basically having some other flag in the ServingRuntime to indicate that it's a "legacy style" mutli-model server, similar to the existing `multiModel` field. Then have a separate ServingRuntime defined for each kind of usage. So in theory you could have three Triton ServingRuntimes defined:
- Single model server
- "Legacy" multi model server
- Modelmesh multi model server

Then if storageUri isn't set we select only runtimes with this flag set. I think this would mean the MLServer-specific logic above in `SetMlServerDefaults()` could also be moved to the runtime definition.

WDYT?

Reply:
For MLServer it requires adding the framework class label based on the model format, if we move to the runtime definition then it requires splitting into multiple runtimes which would be hard to maintain. Maybe the runtime should be like helm template that allows conditions.

Reply:
Created https://github.com/kserve/kserve/issues/2141 for follow up.